### PR TITLE
LF-4221 Alternative Redux handling

### PR DESCRIPTION
**Pull request by kathyavini**
**Description**

For discussion as mentioned in the [comments](https://github.com/LiteFarmOrg/LiteFarm/pull/3294#pullrequestreview-2180799166) of
- #3294 

Summary of changes:
- Let the task product array reside within the task Redux entity, rather than as its own entity that is re-inserted using the task selector
- By considering task products as a property rather than an entity, the Redux store is made less (artificially) flat, but better mirrors the API

Notes:
- The (backend) graph handles the embedding and updating of the task products, so further manipulation is not needed on the frontend.
- The old pattern that task uses creates a flat Redux structure from a returned graph at the expense of a lot of code to manually manage the entities. It has been prone to bugs
- Our newer pattern (enforced by RTK Query) has been to let the backend and Redux entities be parallel

Jira link:

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

**Review by SayakaOno — APPROVED** [review]
Looks good, and works as expected!

File: packages/api/src/controllers/taskController.js [hunk 1]
```diff
@@ -702,6 +702,14 @@ const taskController = {
         )
         .whereIn('task_id', taskIds);
       const filteredTasks = graphTasks.map(removeNullTypes);
+
+      /* Clean before returning to frontend */
+      filteredTasks.forEach((task) => {
+        if (!task.soil_amendment_task) {
```
Comment (Duncan-Brain) [inline packages/api/src/controllers/taskController.js:708]: It is not guaranteed that a soil amendment task will have `soil_amendment_task` property (see the selectId error) . I hope this is the case moving forward but the most reliable way is to check based on type_id.

Comment (kathyavini) [reply packages/api/src/controllers/taskController.js:708]: Could you fix it? 😇 

I'm on the frontend for a while with bugs

Comment (Duncan-Brain) [reply packages/api/src/controllers/taskController.js:708]: @kathyavini branch or commit?

**Review by Duncan-Brain — APPROVED** [review]
Nice simplification!

**Review by SayakaOno — APPROVED** [review]

**Review by SayakaOno — COMMENTED** [review]
The test `should get all tasks for a farm` is failing with the error message `Cannot read properties of undefined (reading 'task_type_id')`

File: packages/api/src/controllers/taskController.js [hunk 2]
```diff
@@ -702,6 +702,20 @@ const taskController = {
...
+        task_type_id: soilAmendmentTypeId,
+      } = await TaskTypeModel.query()
+        .whereNotDeleted()
+        .where({ farm_id: null, task_translation_key: 'SOIL_AMENDMENT_TASK' })
+        .first();
+      filteredTasks.forEach((task) => {
+        if (task.task_type_id !== soilAmendmentTypeId) {
+          delete task.soil_amendment_task_products;
+        }
+      });
```
Comment (SayakaOno) [inline packages/api/src/controllers/taskController.js:717]: We don't have `SOIL_AMENDMENT_TASK` type in the test DB!

Comment (Duncan-Brain) [reply packages/api/src/controllers/taskController.js:717]: Dang tests... quess we missed this on the branch pr haha

Comment (kathyavini) [reply packages/api/src/controllers/taskController.js:717]: Thank you @SayakaOno 🙏

**Review by Duncan-Brain — APPROVED** [review]

**Review by SayakaOno — APPROVED** [review]

**Pull request merged**